### ISSUE TITLE
[Bug]: WebIfc.IFC2X3.IfcDirection uses number instead of IfcReal for DirectionRatios

### ISSUE THREAD

**Issue by jonathan-buildrz** [IFCjs/web-ifc#388]
### What happened?

Contrary to IfcDirection in schema 4 and 4.3, the 2.3 class uses an array of number instead of an array of IfcReal for its DirectionRatios attributes. If I pass numbers into the constructor, I get a bunch of "ERROR: Error in writeline: unknown object received" in the console, and the numbers are not written in the output ifc file.
If instead I pass IfcReals to the constructor (and tell TS to not care about it by using "as unknown as number"), the output is correct.
Here is a buggy output line :
#10=IFCDIRECTION(());
Here is a correct one :
#10=IFCDIRECTION((0.,1.,0.));

### Version

0.0.39

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
ERROR: Error in writeline: unknown object received
```

**Comment by jonathan-buildrz** [comment]
I stumbled upon a few other attributes that don't seem to output correctly (RefLatitude/RefLongitude as IfcCompoundPlaneAngleMeasure in IfcSite ; Orientation as IfcBoolean in IfcFaceBound/IfcFaceOuterBound) ; I wonder if I should open a bug report for each one of them or if they are all related to one root problem ?

**Comment by beachtom** [comment]
Can you share any code for this. I can't seem to reproduce this.

**Comment by jonathan-buildrz** [comment]
@beachtom 
with this._api being a WebIfc.IfcAPI object,
This 2.3 IFC type correct code produces the incorrect IFC line (everything else skipped for brevity), with a message repeated in console 3 times :
ERROR: Error in writeline: unknown object received
```
    const modelID = this._api.CreateModel({schema: 'IFC2X3'});
    await this._api.WriteLine(modelID, new WebIfc.IFC2X3.IfcDirection(1000000, [
      1,
      0,
      0
    ]));
    return this._api.SaveModel(modelID);
```
`#1000000=IFCDIRECTION(());`

Working around the types produces a correct line
```
    const modelID = this._api.CreateModel({schema: 'IFC2X3'});
    await this._api.WriteLine(modelID, new WebIfc.IFC2X3.IfcDirection(1000000, [
      new WebIfc.IFC2X3.IfcReal(1) as unknown as number,
      new WebIfc.IFC2X3.IfcReal(0) as unknown as number,
      new WebIfc.IFC2X3.IfcReal(0) as unknown as number,
    ]));
    return this._api.SaveModel(modelID);
```
`#1000000=IFCDIRECTION((1.,0.,0.));`

IFC4.3 Behaves as expected
```
    const modelID = this._api.CreateModel({schema: 'IFC4_3'});
    await this._api.WriteLine(modelID, new WebIfc.IFC4_3.IfcDirection(1000000, [
      new WebIfc.IFC2X3.IfcReal(1),
      new WebIfc.IFC2X3.IfcReal(0),
      new WebIfc.IFC2X3.IfcReal(0),
    ]));
    return this._api.SaveModel(modelID);
```
`#1000000=IFCDIRECTION((1.,0.,0.));`

For booleans, the following 2.3 type correct code produces an incorrect line
```
    const modelID = this._api.CreateModel({schema: 'IFC2X3'});
    await this._api.WriteLine(modelID, new WebIfc.IFC2X3.IfcFaceBound(1000001,
      new WebIfc.IFC2X3.IfcLoop(1000002),
      true
    ));
    return this._api.SaveModel(modelID);
  }
```
```
#1000002=IFCLOOP();
#1000001=IFCFACEBOUND(#1000002);
```
Working around the types in an even more brutish way allows to produce a correct line (I think I took that from web-ifc-viewer when I was using 0.0.35)
```
    const modelID = this._api.CreateModel({schema: 'IFC2X3'});
    await this._api.WriteLine(modelID, new WebIfc.IFC2X3.IfcFaceBound(1000001,
      new WebIfc.IFC2X3.IfcLoop(1000002),
     {type: 3, value: 'T'} as unknown as boolean
    ));
    return this._api.SaveModel(modelID);
```
```
#1000002=IFCLOOP();
#1000001=IFCFACEBOUND(#1000002,.T.);
```
In IFC 4.3, IfcBoolean seem to provoke an exception, using the following code
```
    const modelID = this._api.CreateModel({schema: 'IFC4_3'});
    await this._api.WriteLine(modelID, new WebIfc.IFC4_3.IfcFaceBound(1000001,
      new WebIfc.IFC4_3.IfcLoop(1000002),
      new WebIfc.IFC4_3.IfcBoolean(true)
    ));
    return this._api.SaveModel(modelID);
```
```
    "name": "BindingError",
    "message": "Cannot pass non-string to std::string",
    "stack": "BindingError: Cannot pass non-string to std::string
    at BindingError.<anonymous> (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:8990:25)
    at new BindingError (eval at createNamedFunction (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:8984:18), <anonymous>:4:34)
    at throwBindingError (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:9182:17)
    at Object.toWireType (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:10198:15)
    at __emval_as (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:10404:42)
    at http://localhost:8080/web-ifc.wasm:wasm-function[639]:0x7e097
    at http://localhost:8080/web-ifc.wasm:wasm-function[404]:0x4a949
    at http://localhost:8080/web-ifc.wasm:wasm-function[975]:0x924c4
    at http://localhost:8080/web-ifc.wasm:wasm-function[931]:0x912b9
    at Object.WriteLine (eval at new_ (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:9810:31), <anonymous>:11:1)
    at IfcAPI2.WriteRawLineData (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:63358:21)
    at IfcAPI2.WriteLine (http://localhost:8080/npm.web-ifc.aa67f0387aa198170254.js:63338:10)
```

**Comment by jonathan-buildrz** [comment]
So I guess we have two problems :
- Some ifc 2.3 objects using raw js types instead of ifc types, where the 4.3 ones are correct
- ifc booleans not working at all for writing

**Comment by beachtom** [comment]
Ok thanks. Can you confirm you are using at least .39? I thought I fixed the bool issue - but if you are then I guess not

**Comment by jonathan-buildrz** [comment]
Yes, I see 0.0.39 in my package.json and IFC header.
I saw the 0.0.40 was published the very day I started upgrading from 36 to 39. So I quickly tested with 40 ; now there is no crash anymore, but the IfcBoolean still isn't written. And the brutish way (passing an object with type and value) doesn't seem to work either.
I'm mostly interested in IFC 2.3 for now, testing with 4.3 was more for the sake of comparison and compleness.
